File: internal/incident/statemachine.go
// Package incident holds the deterministic incident state machine. This is the spine
// of the system: the LLM never mutates it. The orchestrator asks the machine to make
// a transition, the machine checks it's legal, records it, and refuses anything else.
// Keeping this deterministic and separate from the model is what lets an
// AI-in-the-loop system stay auditable.
package incident

import (
	"fmt"
	"time"

	"github.com/parag-labs/incident-commander/pkg/models"
)

// allowed maps each state to the states it may legally move to. A transition not in
// this table is rejected - there is no path the machine will take that isn't declared
// here, so the flow can't be steered somewhere unexpected by model output.
var allowed = map[models.State][]models.State{
	models.StateNew:                 {models.StateTriaging},
	models.StateTriaging:            {models.StateInvestigating, models.StateInvestigationFailed},
	models.StateInvestigating:       {models.StateRootCauseIdentified, models.StateInvestigationFailed},
	models.StateRootCauseIdentified: {models.StateRemediationProposed, models.StateResolved, models.StateEscalated},
	models.StateRemediationProposed: {models.StateWaitingForApproval, models.StateRemediating, models.StateEscalated},
	models.StateWaitingForApproval:  {models.StateRemediating, models.StateEscalated},
	models.StateRemediating:         {models.StateVerifying, models.StateRemediationFailed},
	models.StateVerifying:           {models.StateResolved, models.StateVerificationFailed},

	// Terminal / failure states have no outgoing transitions.
	models.StateResolved:            {},
	models.StateInvestigationFailed: {},
	models.StateRemediationFailed:   {},
	models.StateVerificationFailed:  {},
	models.StateEscalated:           {},
}

// terminal is the set of states from which the incident is finished.
var terminal = map[models.State]bool{
	models.StateResolved:            true,
	models.StateInvestigationFailed: true,
	models.StateRemediationFailed:   true,
	models.StateVerificationFailed:  true,
	models.StateEscalated:           true,
}

// ErrIllegalTransition is returned when a caller asks for a move the machine forbids.
type ErrIllegalTransition struct {
	From models.State
	To   models.State
}

func (e ErrIllegalTransition) Error() string {
	return fmt.Sprintf("illegal transition %s -> %s", e.From, e.To)
}

// CanTransition reports whether from -> to is a declared, legal move.
func CanTransition(from, to models.State) bool {
	for _, s := range allowed[from] {
		if s == to {
			return true
		}
	}
	return false
}

// IsTerminal reports whether an incident in this state is finished.
func IsTerminal(s models.State) bool { return terminal[s] }

// New creates a fresh incident in the NEW state from an alert.
func New(id string, alert models.Alert, now time.Time) *models.Incident {
	title := fmt.Sprintf("%s on %s", alert.Metric, alert.Service)
	return &models.Incident{
		ID:        id,
		Title:     title,
		State:     models.StateNew,
		Alert:     alert,
		CreatedAt: now,
		UpdatedAt: now,
		History: []models.Transition{
			{From: "", To: models.StateNew, Reason: "incident opened", At: now},
		},
	}
}

// Transition advances the incident to a new state if the move is legal, recording it
// in the history. It returns ErrIllegalTransition and leaves the incident untouched
// otherwise - the machine never half-applies a move.
func Transition(inc *models.Incident, to models.State, reason string, now time.Time) error {
	if !CanTransition(inc.State, to) {
		return ErrIllegalTransition{From: inc.State, To: to}
	}
	inc.History = append(inc.History, models.Transition{
		From: inc.State, To: to, Reason: reason, At: now,
	})
	inc.State = to
	inc.UpdatedAt = now
	return nil
}

File: java/src/main/java/com/incidentcommander/StateMachine.java
package com.incidentcommander;

import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * The deterministic incident lifecycle.
 *
 * <p>This is the spine of the system: the LLM never mutates it. The orchestrator asks the
 * machine to make a transition, the machine checks it is legal, records it, and refuses
 * anything else.
 */
public final class StateMachine {
    private StateMachine() {
    }

    // ALLOWED maps each state to the states it may legally move to. A transition not in
    // this table is rejected - there is no path the machine will take that is not declared
    // here, so the flow cannot be steered somewhere unexpected by model output.
    private static final Map<String, List<String>> ALLOWED = Map.ofEntries(
            Map.entry(States.NEW, List.of(States.TRIAGING)),
            Map.entry(States.TRIAGING, List.of(States.INVESTIGATING, States.INVESTIGATION_FAILED)),
            Map.entry(
                    States.INVESTIGATING,
                    List.of(States.ROOT_CAUSE_IDENTIFIED, States.INVESTIGATION_FAILED)),
            Map.entry(
                    States.ROOT_CAUSE_IDENTIFIED,
                    List.of(States.REMEDIATION_PROPOSED, States.RESOLVED, States.ESCALATED)),
            Map.entry(
                    States.REMEDIATION_PROPOSED,
                    List.of(States.WAITING_FOR_APPROVAL, States.REMEDIATING, States.ESCALATED)),
            Map.entry(
                    States.WAITING_FOR_APPROVAL,
                    List.of(States.REMEDIATING, States.ESCALATED)),
            Map.entry(States.REMEDIATING, List.of(States.VERIFYING, States.REMEDIATION_FAILED)),
            Map.entry(States.VERIFYING, List.of(States.RESOLVED, States.VERIFICATION_FAILED)),
            Map.entry(States.RESOLVED, List.of()),
            Map.entry(States.INVESTIGATION_FAILED, List.of()),
            Map.entry(States.REMEDIATION_FAILED, List.of()),
            Map.entry(States.VERIFICATION_FAILED, List.of()),
            Map.entry(States.ESCALATED, List.of()));

    // TERMINAL is the set of states from which the incident is finished.
    private static final Set<String> TERMINAL = Set.of(
            States.RESOLVED,
            States.INVESTIGATION_FAILED,
            States.REMEDIATION_FAILED,
            States.VERIFICATION_FAILED,
            States.ESCALATED);

    /**
     * Report whether {@code from -> to} is a declared, legal move.
     *
     * @param from the source state
     * @param to the destination state
     * @return whether the move is legal
     */
    public static boolean canTransition(String from, String to) {
        return ALLOWED.getOrDefault(from, List.of()).contains(to);
    }

    /**
     * Report whether an incident in this state is finished.
     *
     * @param state the state to test
     * @return whether the state is terminal
     */
    public static boolean isTerminal(String state) {
        return TERMINAL.contains(state);
    }

    /**
     * Create a fresh incident in the {@code NEW} state from an alert.
     *
     * @param id the incident identifier
     * @param alert the alert that opened it
     * @param now the creation timestamp
     * @return the new incident
     */
    public static Incident newIncident(String id, Alert alert, Instant now) {
        Incident inc = new Incident();
        inc.id = id;
        inc.title = alert.metric + " on " + alert.service;
        inc.state = States.NEW;
        inc.alert = alert;
        inc.createdAt = now;
        inc.updatedAt = now;
        Transition opening = new Transition();
        opening.from = "";
        opening.to = States.NEW;
        opening.reason = "incident opened";
        opening.at = now;
        inc.history.add(opening);
        return inc;
    }

    /**
     * Advance the incident to a new state if the move is legal, recording it. Throws
     * {@link IllegalTransitionException} and leaves the incident untouched otherwise - the
     * machine never half-applies a move.
     *
     * @param inc the incident to advance
     * @param to the destination state
     * @param reason why the move happens
     * @param now the transition timestamp
     */
    public static void transition(Incident inc, String to, String reason, Instant now) {
        if (!canTransition(inc.state, to)) {
            throw new IllegalTransitionException(inc.state, to);
        }
        Transition step = new Transition();
        step.from = inc.state;
        step.to = to;
        step.reason = reason;
        step.at = now;
        inc.history.add(step);
        inc.state = to;
        inc.updatedAt = now;
    }
}
